### PR TITLE
feat: batch Discord alerts for pollDecisions cron

### PR DESCRIPTION
**Pull request by kouloumos**
## Overview
The `pollDecisions` cron job dispatches up to 10 tasks per run. Each task was sending individual "started" and "completed" Discord alerts — producing up to 20 messages per invocation, most of which reported no new decisions.

This PR replaces per-task alerts with batch summaries:
- **1 batch started alert** when the cron dispatches tasks (counts, meeting list, any dispatch errors)
- **1 batch completed alert** when the last task in the batch finishes (aggregated results, per-meeting breakdown, admin links)

Net effect: **20 messages → 2 messages** per cron run. Manual triggers from the admin panel or subject page still get individual started alerts plus an immediate completion summary.

## How it works

- **`DiscordAlertMode`** (`types.ts`): New config on `TASK_CONFIG` — `'none'` suppresses all generic Discord alerts for a task type. Only `pollDecisions` uses it; all others default to `'all'`.
- **`silent` option** (`tasks.ts`): `startTask()` accepts `silent: true` to skip the "started" alert. The cron batch uses this while manual triggers don't.
- **Batch completion detection** (`pollDecisions.ts`): After processing a result, `checkBatchCompletionAndNotify()` queries sibling `pollDecisions` tasks created within a ±2 minute window. When all are in a terminal state, it aggregates results and sends a single summary. Runs in a `finally` block so failures aren't silently lost.

## Test plan

- Trigger cron endpoint and verify Discord channel gets 2 messages instead of 20
- Trigger manual poll from admin panel and verify individual started alert + immediate completion summary

---

> [!NOTE]
> **Medium Risk**
> Moderate risk because it changes task lifecycle alerting/observability and adds new terminal-hook execution after task updates, which could impact when/if Discord notifications fire and how `TaskStatus.responseBody` is written.
> 
> **Overview**
> Reduces Discord noise from `pollDecisions` by **suppressing generic per-task started/completed/failed alerts** and replacing them with **batch-level cron alerts** (one dispatch summary + one aggregated completion summary).
> 
> Adds `DiscordAlertMode` to `TASK_CONFIG` (with `pollDecisions` set to `none`), a `silent` option on `startTask`, and a new `taskTerminalHooks` mechanism so `pollDecisions` can detect when all sibling tasks in a ±2 minute window are terminal and then send a single completion embed with counts, per-meeting details, admin links, and safe field truncation.
> 
> Updates `pollDecisions` processing to enrich stored `responseBody` with post-processing counts for accurate aggregation, expands test coverage around alert gating and batching, and documents the new alert behavior in `docs/admin-alerts.md`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3f52a38b9c41bee569c6736ee5415003658ff7f0. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h3>Greptile Summary</h3>

This PR reduces Discord alert noise for the `pollDecisions` cron job from up to 20 per-task messages down to 2 batch-level messages — one started summary at dispatch time and one aggregated completion summary when all sibling tasks finish. It introduces `DiscordAlertMode` configuration in `TASK_CONFIG`, a `silent` option on `startTask`, and a `taskTerminalHooks` mechanism that fires after `handleTaskUpdate` has fully settled a task's DB status, eliminating the previously noted stale-status race condition.

Key changes:
- **`types.ts`**: New `DiscordAlertMode` type (`'all' | 'none'`), explicit `TaskConfig` interface, and `getDiscordAlertMode()` helper. `pollDecisions` is the only task type set to `'none'`.
- **`tasks.ts`**: `startTask` gains a `silent` flag; `handleTaskUpdate` gates all `sendTaskAdminAlert` calls behind `sendGenericAlerts`; a `taskTerminalHooks` dispatch runs at the very end of `handleTaskUpdate` after all DB writes are committed.
- **`pollDecisions.ts`**: Cron dispatch tracks `skipped`/`dispatchedMeetings`/`dispatchErrors` and fires a single `sendPollDecisionsBatchStartedAlert`. `checkBatchCompletionAndAlert` (the terminal hook) queries all `pollDecisions` sibling tasks in a ±2-minute window; when all are terminal it aggregates results (using `_processedCounts` enriched into `responseBody` by `handlePollDecisionsResult`) and fires `sendPollDecisionsBatchCompletedAlert`.
- **`discord.ts`**: Two new embed functions with field truncation via `truncateField()`, failure-aware title logic for the completed alert, and `.catch()` wrappers on both fire-and-forget calls.
- Two new test suites provide thorough coverage of the `silent` flag, `discordAlertMode` gating, terminal hook lifecycle, and batch aggregation logic.

<h3>Confidence Score: 4/5</h3>

- PR is safe to merge; two minor style improvements are worth addressing but neither causes data loss or silent failures.
- The core logic is well-engineered: the terminal hook correctly fires after DB status is settled (fixing the previously noted race), the `_processedCounts` enrichment uses a try/catch so failure can't mark a succeeded task as failed, and field truncation is applied consistently. Two residual issues keep the score below 5: (1) the started alert uses a fixed blue color even when all tasks fail at dispatch — since that's the only alert fired in a complete-dispatch-failure scenario, the color matters; (2) `cityId` lacks the null-safety fallback (`?? 'unknown'`) that `councilMeetingId` already has. Test coverage is comprehensive and directly validates the new mechanisms.
- `src/lib/discord.ts` (started alert color), `src/lib/tasks/pollDecisions.ts` (cityId null guard in `checkBatchCompletionAndAlert`)

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/lib/tasks/tasks.ts | Adds `silent` option to `startTask`, introduces `getDiscordAlertMode` gating throughout `handleTaskUpdate`, and appends the `taskTerminalHooks` dispatch at the end. Terminal hook correctly fires after all DB writes are settled (both `success` and `error` paths). Logic is sound. |
| src/lib/tasks/pollDecisions.ts | Adds batch dispatch tracking and `checkBatchCompletionAndAlert`. Minor: `cityId` lacks a null-safety fallback (unlike `councilMeetingId`). Counter variables correctly moved outside the transaction so enrichment can access post-processing totals. Documented race conditions (duplicate alerts, manual-trigger grouping) are acceptable trade-offs. |
| src/lib/discord.ts | Adds `sendPollDecisionsBatchStartedAlert` and `sendPollDecisionsBatchCompletedAlert` with proper field truncation via `truncateField`. The started alert uses a fixed blue color even when dispatch errors are present — important because when all tasks fail at dispatch, only this alert fires. Completed alert title correctly differentiates all four state combinations. |
| src/lib/tasks/registry.ts | Cleanly introduces `taskTerminalHooks` as a `Partial<Record<string, TaskTerminalHook>>`, wires `checkBatchCompletionAndAlert` for `pollDecisions`. Type and structure look correct. |
| src/lib/tasks/types.ts | Adds `DiscordAlertMode` type, explicit `TaskConfig` interface, and `getDiscordAlertMode` helper. Correctly defaults unknown task types to `'all'` to prevent accidental alert suppression. Switches from `as const` to `satisfies` for better type checking. |
| src/lib/tasks/__tests__/checkTaskIdempotency.test.ts | Adds comprehensive coverage for the `silent` option and `taskTerminalHooks` mechanism. The two-phase mock (`mockResolvedValueOnce` for succeeded then failed) correctly models the actual flow in `handleTaskUpdate`. Edge cases (hook throws, processing status, non-hooked task types) are all covered. |
| src/lib/tasks/__tests__/pollDecisionsBatchAlerts.test.ts | New test file with thorough coverage of `checkBatchCompletionAndAlert`: all-terminal, partial, single-task, failure mixing, conflict aggregation, `_processedCounts` fallback, malformed responseBody, and error truncation. Well-structured and complete. |
| docs/admin-alerts.md | Documentation is accurate and complete. Correctly documents the 2-message-per-cron-run behaviour, alert format examples, and the BATCH_WINDOW_MS coupling constraint with the cron interval. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Cron as pollDecisionsForRecentMeetings
    participant ST as startTask
    participant Discord as Discord
    participant CB as Callback Endpoint
    participant HTU as handleTaskUpdate
    participant HPR as handlePollDecisionsResult
    participant Hook as checkBatchCompletionAndAlert

    Cron->>ST: pollDecisionsForMeeting(silent:true) ×N
    ST->>ST: DB create task (status=pending)
    ST->>ST: Call backend API
    ST-->>Cron: task created (no started alert)
    Cron->>Discord: sendPollDecisionsBatchStartedAlert(N tasks)

    note over CB,Hook: Each task completes asynchronously

    CB->>HTU: handleTaskUpdate(taskId, status=success)
    HTU->>HTU: DB update → succeeded (raw responseBody)
    HTU->>HPR: processResult (handlePollDecisionsResult)
    HPR->>HPR: $transaction (process decisions)
    HPR->>HPR: DB update responseBody += _processedCounts
    HPR-->>HTU: returns
    HTU->>Hook: taskTerminalHooks[pollDecisions](taskId, createdAt)
    Hook->>Hook: DB query: all siblings in ±2min window
    alt not all terminal
        Hook-->>HTU: return (wait for siblings)
    else all terminal
        Hook->>Hook: aggregate counts from _processedCounts
        Hook->>Discord: sendPollDecisionsBatchCompletedAlert
    end
```

<details><summary><h3>Comments Outside Diff (3)</h3></summary>

1. `src/lib/discord.ts`, line 773-774 ([link](https://github.com/schemalabz/opencouncil/blob/7009a9c805c8fd02402a1d104ba3d2613b835d8d/src/lib/discord.ts#L773-L774)) 

   **`truncateField` not applied to pre-existing `sendTranscriptSendFailedAdminAlert`**

   This PR introduces `truncateField()` as the canonical way to handle Discord's 1024-character field limit (with a visible `… (truncated)` indicator). The `Error` field in `sendTranscriptSendFailedAdminAlert` still uses a raw `.substring(0, 1024)` with a comment instead:

   ```ts
   value: data.error.substring(0, 1024), // Discord field value limit
   ```

   Since the helper is now available, this is an easy clean-up to make the truncation behaviour consistent across all alert functions:

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/discord.ts
   Line: 773-774

   Comment:
   **`truncateField` not applied to pre-existing `sendTranscriptSendFailedAdminAlert`**

   This PR introduces `truncateField()` as the canonical way to handle Discord's 1024-character field limit (with a visible `… (truncated)` indicator). The `Error` field in `sendTranscriptSendFailedAdminAlert` still uses a raw `.substring(0, 1024)` with a comment instead:

   ```ts
   value: data.error.substring(0, 1024), // Discord field value limit
   ```

   Since the helper is now available, this is an easy clean-up to make the truncation behaviour consistent across all alert functions:

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <sub>Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!</sub>

2. `src/lib/discord.ts`, line 129-131 ([link](https://github.com/schemalabz/opencouncil/blob/3f52a38b9c41bee569c6736ee5415003658ff7f0/src/lib/discord.ts#L129-L131)) 

   **Started alert stays blue when dispatch errors occur**

   `sendPollDecisionsBatchStartedAlert` uses a fixed blue color (`0x3498db`) even when `data.errors.length > 0`. The `sendPollDecisionsBatchCompletedAlert` function correctly adjusts color based on failure state, but the started alert does not.

   This is particularly important for the documented edge case where all tasks fail at dispatch time — only the started alert fires in that scenario, so its color is the sole visual signal that something went wrong. An operator glancing at the Discord channel would see a blue embed (indicating normal operation) rather than the expected red.

3. `src/lib/tasks/pollDecisions.ts`, line 1063-1066 ([link](https://github.com/schemalabz/opencouncil/blob/3f52a38b9c41bee569c6736ee5415003658ff7f0/src/lib/tasks/pollDecisions.ts#L1063-L1066)) 

   **`cityId` lacks a null-safety fallback unlike `councilMeetingId`**

   `councilMeetingId` is defensively handled with `?? 'unknown'`, but `cityId` is used directly without a fallback. If the Prisma schema permits `TaskStatus.cityId` to be null (or if it ever becomes nullable in a future migration), this would propagate a `null` into `PollDecisionsMeetingResult.cityId` — which is typed as `string` — causing a type error and potentially malformed `meetingAdminUrl` calls in `discord.ts` (producing links like `null/meeting-id`).

   Consider adding a consistent fallback:

</details>

<sub>Last reviewed commit: 3f52a38</sub>

### CODE REVIEW

File: src/lib/tasks/pollDecisions.ts [hunk 1]
```diff
@@ -806,143 +828,284 @@ export async function resolveAdaConflict(
...
-    // Collect all subjectIds from matches and reassignments for validation
-    const allSubjectIds = [
-        ...result.matches.map(m => m.subjectId),
-        ...result.reassignments.flatMap(r => [r.fromSubjectId, r.toSubjectId]),
-    ];
+/**
+ * Radius (ms) for grouping pollDecisions tasks into a batch.
+ * Tasks created within ±BATCH_WINDOW_MS of each other are considered siblings.
+ */
+const BATCH_WINDOW_MS = 2 * 60 * 1000;
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil/compare/9947e74c7ad1a3a9368df2f771a61cf2ed0bc0c7..968e4114a1fca1dfe161a1d4b2bf3cfc56e2edd4) — added a comment documenting the cron interval coupling.

File: src/lib/tasks/pollDecisions.ts [hunk 2]
```diff
@@ -806,143 +828,284 @@ export async function resolveAdaConflict(
...
+    // Enrich responseBody with post-processing counts so the batch completion
+    // hook (checkBatchCompletionAndAlert) can read accurate totals from DB.
+    await prisma.taskStatus.update({
+        where: { id: taskId },
+        data: {
+            responseBody: JSON.stringify({
+                ...result,
+                _processedCounts: { matches: processedCount, reassignments: reassignmentCount, conflicts: conflictCount },
+            }),
+        },
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil/compare/9947e74c7ad1a3a9368df2f771a61cf2ed0bc0c7..968e4114a1fca1dfe161a1d4b2bf3cfc56e2edd4) — wrapped the post-transaction `responseBody` enrichment in try/catch so a transient DB error is logged but does not propagate and mark the task as failed.

**Comment by kouloumos** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/968e4114a1fca1dfe161a1d4b2bf3cfc56e2edd4..7009a9c805c8fd02402a1d104ba3d2613b835d8d) to address bot review comments ([1](https://github.com/schemalabz/opencouncil/pull/255#discussion_r2901648032), [2](https://github.com/schemalabz/opencouncil/pull/255#discussion_r2901652376), [3](https://github.com/schemalabz/opencouncil/pull/255#discussion_r2901652377)):

- Batch completion title now surfaces failures: "N failure(s), no decisions found" instead of misleading "no new decisions"
- Added top-level "Conflicts" summary field alongside "Reassignments" for consistency
- Fixed indentation in `handlePollDecisionsResult` — dedented refactoring artifact from 8 to 4 spaces

File: src/lib/tasks/tasks.ts [hunk 3]
```diff
@@ -163,16 +163,18 @@ export const startTask = async (taskType: MeetingTaskType, requestBody: any, cou
...
+        sendTaskAdminAlert({
+            status: 'started',
+            taskType: taskType,
+            cityName: newTask.councilMeeting.city.name_en,
+            meetingName: newTask.councilMeeting.name_en,
+            taskId: newTask.id,
+            cityId: cityId,
+            meetingId: councilMeetingId,
+        });
+    }
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/tasks.ts]: Valid observation — `discordAlertMode: 'none'` does suppress all generic lifecycle alerts for `pollDecisions`, making `silent` redundant for this task type. The PR description is slightly inaccurate: manual triggers don't get a separate "started" alert, they only get the batch completion summary (as a batch of 1). Will update the PR description. The `silent` flag itself remains useful as a general-purpose mechanism for other task types.

File: src/lib/discord.ts [hunk 4]
```diff
@@ -604,6 +605,146 @@ export async function sendTranscriptSentAdminAlert(data: {
...
+        });
+    }
+
+    await sendAdminAlert({
+        title,
+        description: `${data.succeededCount + data.failedCount} task(s) completed`,
+        color,
+        fields,
+    });
+}
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/discord.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil/compare/968e4114a1fca1dfe161a1d4b2bf3cfc56e2edd4..7009a9c805c8fd02402a1d104ba3d2613b835d8d) — added a top-level "Conflicts" summary field, consistent with the existing "Reassignments" field.

File: src/lib/tasks/pollDecisions.ts [hunk 5]
```diff
@@ -806,6 +830,145 @@ export async function resolveAdaConflict(
...
+ * 'failed' directly without a callback, so handleTaskUpdate and this hook are
+ * never invoked for them. Partial dispatch failures self-heal (surviving tasks
+ * trigger this hook and find failed siblings via the time-window query). A
+ * complete dispatch failure produces only a batch started alert — no completed
+ * alert fires, since the started alert already shows the dispatch errors.
+ */
+export async function checkBatchCompletionAndAlert(
+    taskId: string,
+    taskCreatedAt: Date,
+) {
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil/compare/e99ebb618c685b414731c13ae5d4c3d73ce0a300..3f52a38b9c41bee569c6736ee5415003658ff7f0) — prefixed with underscore.
